R/numberize.R

```diff
@@ -0,0 +1,171 @@
+  digits <- numbers[match(words, numbers[[lang]]), "digit"]
+  digits
+}
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#' Generate a number from a numeric vector.
+#' Uses `digits_from()` output to generate the numeric value of the text.
+#'
+#' @param digits A numeric vector translated from words.
+#'
+#' @return A numeric value.
+#'
+#' @keywords internal
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+number_from <- function(digits) {
+  thousand_index <- match(1000, digits, nomatch = 0)
+  million_index <- match(1E6, digits, nomatch = 0)
+
+  # for lang = "es" multiply 1000 * 1E6 for billion
+  if (thousand_index < million_index) { # es thousand million = billion
+    digits[thousand_index] <- digits[thousand_index] * digits[million_index]
+  }
+
+  summed <- 0
+  total <- 0
+  for (d in digits) {
+    if (d %in% c(1E3, 1E6, 1E9, 1E12)) {
+      total <- ifelse(summed == 0, total + d, total + summed * d)
+      summed <- 0
+    } else if (d == 100) {
+      summed <- ifelse(summed == 0, d, summed * d)
+    } else {
+      summed <- summed + d
+    }
+  }
+  summed + total
+}
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#' Internal function used in the numberize() call for vectors.
+#'
+#' @param text Character string in a supported language.
+#' @param lang Language of the character string.
+#' Currently one of `"en" | "fr" | "es"`.
+#'
+#' @return A numeric value.
+#'
+#' @keywords internal
+#'
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+.numberize <- function(text, lang = c("en", "fr", "es")) {
+  # return NA if the input is NA
+  if (is.na(text)) {
+    return(NA)
+  }
+
+  # convert to numeric. Numeric values will pass and non numeric values will be
+  # coerced to NA and converted into numbers.
+  tmp_text <- suppressWarnings(as.numeric(text))
+  if (!is.na(tmp_text)) {
+    return(tmp_text)
+  } else {
+    # when the text does not correspond to a number, digits_from() returns NA
+    digits <- digits_from(text, lang)
+    if (anyNA(digits)) {
+      return(NA)
+    }
+    number_from(digits)
+  }
+}
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#' Convert a vector string of spelled numbers in a supported language to
+#' its numeric equivalent.
+#'
+#' @param text Vector containing spelled numbers in a supported language.
+#' @param lang The text's language. Currently one of `"en" | "fr" | "es"`.
+#' Default is "en"
+#'
+#' @return A vector of numeric values.
+#'
+#' @examples
+#' # convert to numbers a scalar
+#' numberize("five hundred and thirty eight")
+#'
+#' # convert a vector of values
+#' numberize(c("dix", "soixante-cinq", "deux mille vingt-quatre"), lang = "fr")
+#'
+#' @export
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+numberize <- function(text, lang = c("en", "fr", "es")) {
+  lang <- match.arg(lang)
```

Comment:
```suggestion
numberize <- function(text, lang = c("en", "fr", "es")) {
  lang <- tolower(lang)
  lang <- match.arg(lang)
```

This hardens the input for the `lang` argument - it can happen that `EN` is entered rather easily.

Reply:
Noted and modified accordingly.